crates/net/eth-wire/src/muxdemux.rs

```diff
@@ -0,0 +1,391 @@
+    }
+
+    fn shared_cap<M: 'static>(
+        shared_capabilities: &SharedCapabilities,
+    ) -> Result<SharedCapability, MuxDemuxError> {
+        for shared_cap in shared_capabilities.iter_caps() {
+            match shared_cap {
+                SharedCapability::Eth { .. } if TypeId::of::<M>() == TypeId::of::<EthMessage>() => {
+                    return Ok(shared_cap)
+                }
+                _ => continue,
+                // more caps to come ..
+            }
+        }
+
+        Err(CapabilityNotShared)
+    }
+
+    fn reg_new_ingress_buffer(
+        &mut self,
+        cap: SharedCapability,
+    ) -> Result<mpsc::UnboundedReceiver<BytesMut>, MuxDemuxError> {
+        if let Some(tx) = self.demux.get(&cap) {
+            if !tx.is_closed() {
+                return Err(StreamAlreadyExists)
+            }
+        }
+        let (ingress_tx, ingress) = mpsc::unbounded_channel();
+        self.demux.insert(cap, ingress_tx);
+
+        Ok(ingress)
+    }
+
+    fn unmask_msg_id(&self, id: &mut u8) -> Result<SharedCapability, MuxDemuxError> {
+        for cap in self.shared_capabilities.iter_caps() {
+            let offset = cap.relative_message_id_offset();
+            let next_offset = offset + cap.num_messages()?;
+            if *id < next_offset {
+                *id -= offset;
+
+                return Ok(cap)
+            }
+        }
+
+        Err(MessageIdOutOfRange(*id))
+    }
+
+    /// Masks message id with offset relative to the message id suffix reserved for capability
+    /// message ids. The p2p stream further masks the message id (todo: mask whole message id at
+    /// once to avoid copying message to mutate id byte or sink BytesMut).
+    fn mask_msg_id(&self, msg: Bytes) -> Bytes {
+        let mut masked_bytes = BytesMut::zeroed(msg.len());
+        masked_bytes[0] = msg[0] + self.owner.relative_message_id_offset();
+        masked_bytes[1..].copy_from_slice(&msg[1..]);
+
+        masked_bytes.freeze()
+    }
+
+    /// Checks if all clones of this shared stream have been dropped, if true then returns //
+    /// function to drop the stream.
+    fn can_drop(&mut self) -> bool {
+        for tx in self.demux.values() {
+            if !tx.is_closed() {
+                return false
+            }
+        }
+
+        true
+    }
+}
+
+impl<S, E> Stream for MuxDemuxStream<S>
+where
+    S: Stream<Item = Result<BytesMut, E>> + CanDisconnect<Bytes> + Unpin,
+    MuxDemuxError: From<E> + From<<S as Sink<Bytes>>::Error>,
+{
+    type Item = Result<BytesMut, MuxDemuxError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // sink buffered bytes from `StreamClone`s
+        for _ in 0..self.demux.len() {
+            let Ok(item) = self.mux.try_recv() else { break };
+            if self.poll_ready_unpin(cx).is_pending() {
+                break
+            }
+            self.inner.start_send_unpin(item)?;
+        }
```

Comment:
this can lead to dropped messages.
we can either use an internal vecdeque to buffer the messages pulled from the streams (should be fine because the channels itself are unbounded so doesn't really matter where we buffer)

but definitely put this in a loop over poll_ready

Reply:
true, the `is_pending check` should go first